lib/puppet/indirector/file_bucket_file/file.rb

```diff
@@ -189,19 +210,45 @@ def path_for(bucket_path, digest, subfile = nil)
       Puppet::FileSystem.pathname(subfile ? ::File.join(basedir, subfile) : basedir)
     end
 
-    # @param contents_file [Object] Opaque file path
-    # @param bucket_file [IO]
-    def verify_identical_file!(contents_file, bucket_file)
-      if bucket_file.size == Puppet::FileSystem.size(contents_file)
-        if bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) }
-          Puppet.info "FileBucket got a duplicate file #{bucket_file.checksum}"
-          return
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [Boolean] whether the data in contents_file is of the same size
+    #   and content as that in the bucket_file
+    # @api private
+    def verify_identical_file(contents_file, bucket_file)
+      (bucket_file.size == Puppet::FileSystem.size(contents_file)) &&
+        (bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) })
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param expected_checksum_data [String] expected value of checksum of type
+    #   checksum_type
+    # @param checksum_type [String] type of check sum of checksum_data, ie "md5"
+    # @return [Boolean] whether the checksum of the contents_file matches the
+    #   supplied checksum
+    # @api private
+    def contents_file_matches_checksum?(contents_file, expected_checksum_data, checksum_type)
+      contents_file_checksum_data = Puppet::Util::Checksums.method(:"#{checksum_type}_file").call(contents_file.to_path)
+      contents_file_checksum_data == expected_checksum_data
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [void]
+    # @api private
+    def copy_bucket_file_to_contents_file(contents_file, bucket_file)
+      Puppet::Util.replace_file(contents_file, 0440) do |of|
```

Comment:
In this `Puppet::Util::replace_file` implementation, we no longer specify `wb` as our open mode option for the contents file. I think we should be OK as to whether we are still reading and writing files correctly in binary encoding, but I'm noting the thinking here for posterity.

`bucket_file.stream` here will either be doing a `FileContents#stream` or a `StringContents#stream`. `FileContents#stream` opens the file for reading in binary: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_bucket/file.rb#L128. `StringContents#stream` creates a StringIO object which preserves the encoding of the original string content (which I think is likely UTF-8). https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_bucket/file.rb#L98.

When we move to file writing inside the block, we're using `FileUtils.copy_stream`.. which is effectively an alias to `IO::copy_stream`. `IO::copy_stream` docs don't explicitly state how it handles encoding (https://ruby-doc.org/core-2.3.0/IO.html#method-c-copy_stream). However, [this bug](https://bugs.ruby-lang.org/issues/8767) implies that `IO::copy_stream` writes in binary mode. That fix was introduced in ruby 2.1.0 and backported to ruby 1.9.3. So I believe we're still good on encoding.